### PR TITLE
Improve assertion re. appliers reading applied state

### PR DESCRIPTION
**Pull request by DaveCTurner**
Calling `ClusterApplierService#state()` from within a cluster applier is
almost certainly a bug, and trips an assertion, but the explanatory
message is not very clear. This commit improves the message.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierServiceTests.java [hunk 1]
```diff
@@ -351,9 +351,14 @@ public void testClusterStateApplierCantSampleClusterState() throws InterruptedEx
                 clusterApplierService.state();
                 error.set(new AssertionError("successfully sampled state"));
             } catch (AssertionError e) {
-                if (e.getMessage().contains("should not be called by a cluster state applier") == false) {
-                    error.set(e);
+                final var message = e.getMessage();
+                if (message.contains("On the cluster applier thread")
+                    && message.contains("you must use ClusterChangedEvent#state()")
+                    && message.contains("almost certainly a bug to read the latest-applied state")
+                    && message.contains("the new state has been committed at this point but is not yet applied")) {
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierServiceTests.java:358]: It is surprising to see only parts of the message, is there a reason to check it this way?

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierServiceTests.java:358]: I feel bad copy-pasting a literal string from the production code to the test. Normally I do this to highlight the important parts in the test, although in this case most of the string is important.

Comment (idegtiarenko) [reply server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierServiceTests.java:358]: Is it worth introducing a constant with error message?

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierServiceTests.java:358]: That has the same problem, it makes it easy to change the message content in a way that loses some information we currently consider to be important.

**Review by idegtiarenko — APPROVED** [review]

**Review by pxsalehi — APPROVED** [review]

File: server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierServiceTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierServiceTests.java]: Ok, I've done that in 94c3b282506.

**Pull request merged**